infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/jwtproxy/JwtProxyProvisioner.java

```diff
@@ -0,0 +1,210 @@
+import io.fabric8.kubernetes.api.model.VolumeBuilder;
+import io.fabric8.kubernetes.api.model.VolumeMount;
+import java.security.KeyPair;
+import java.util.HashMap;
+import java.util.Map;
+import org.eclipse.che.api.core.model.workspace.config.MachineConfig;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalInfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalMachineConfig;
+import org.eclipse.che.multiuser.machine.authentication.server.signature.SignatureKeyManager;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.ServerServiceBuilder;
+
+/** @author Sergii Leshchenko */
+public class JwtProxyProvisioner {
+
+  static final int FIRST_AVAILABLE_PORT = 4400;
+
+  static final int JWT_PROXY_MEMORY_LIMIT_BYTES = 128 * 1024 * 1024; // 128mb
+
+  static final String PUBLIC_KEY_HEADER = "-----BEGIN PUBLIC KEY-----\n";
+  static final String PUBLIC_KEY_FOOTER = "\n-----END PUBLIC KEY-----";
+
+  static final String JWTPROXY_IMAGE = "ksmster/jwtproxy";
+  static final String JWT_PROXY_CONFIG_FILE = "config.yaml";
+  static final String JWT_PROXY_MACHINE_NAME = "jwtproxy";
+
+  static final String JWT_PROXY_CONFIG_FOLDER = "/config";
+  static final String JWT_PROXY_PUBLIC_KEY_FILE = "mykey.pub";
+
+  private final SignatureKeyManager signatureKeyManager;
+
+  private final RuntimeIdentity identity;
+
+  private final JwtProxyConfigBuilder proxyConfigBuilder;
+
+  private final String serviceName;
+  private int availablePort;
+
+  public JwtProxyProvisioner(RuntimeIdentity identity, SignatureKeyManager signatureKeyManager) {
+    this.signatureKeyManager = signatureKeyManager;
+
+    this.identity = identity;
+
+    this.proxyConfigBuilder = new JwtProxyConfigBuilder(identity.getWorkspaceId());
+
+    this.serviceName = generate(SERVER_PREFIX, SERVER_UNIQUE_PART_SIZE) + "-jwtproxy";
+    this.availablePort = FIRST_AVAILABLE_PORT;
+  }
+
+  /**
+   * Modifies Kubernetes environment to expose the specified service port via JWTProxy.
+   *
+   * @param k8sEnv Kubernetes environment to modify
+   * @param backendServiceName service name that will be exposed
+   * @param backendServicePort service port that will be exposed
+   * @param protocol protocol that will be used for exposed port
+   * @return JWTProxy service port that expose the specified one
+   * @throws InfrastructureException if any exception occurs during port exposing
+   */
+  public ServicePort expose(
+      KubernetesEnvironment k8sEnv,
+      String backendServiceName,
+      int backendServicePort,
+      String protocol)
+      throws InfrastructureException {
+    ensureJwtProxyInjected(k8sEnv);
+
+    int listenPort = availablePort++;
```

Comment:
Didn't get some use-cases. Would you be willing to explain them to me?
1. What if 2 workspace masters are running at the same time?
1. What if workspaces get deployed to the same single k8s namespace?
1. Is there any guarantee that this class won't be created multiple times since it is not a singleton and is not created automatically by DI container?

Reply:
> 1. What if 2 workspace masters are running at the same time?

There is no issue with that because of a restriction that only 1 Che Workspace Master is able to start a particular workspace at the same time.

> 2. What if workspaces get deployed to the same single k8s namespace?

There are no any issues with that while it is a port of a jwtproxy pod for a particular workspace. Jwtproxy pod will have a unique name within k8s namespace that is set by `UniqueNamesProvisioner` class.

> 3. Is there any guarantee that this class won't be created multiple times since it is not a singleton and is not created automatically by DI container?

Actually, this class MUST BE created once for one `STARTING` workspace, I don't see any possible ways in which we can guarantee that class is used correctly. What we can do - describe in java docs how it should be used.

One more thing that we could implement to prevent conflicts while incorrect usage of this class - use unique objects name - like machine name, pod name, etc. Then if this class will be created two or more times then the only thing that happens - there are more jwtproxy pods that are needed but not ports, objects conflicts should be there and everything would work as expected.
But I think it may be a bit overkill while this class is supposed to use in one place.

One more note, about unique names - do you think should we use a unique machine name instead of hardcoded? Now we always use `jwtproxy` machine name. It would be better to use generated name with prefix like `jwtproxy9sjXnw6y` to avoid conflicts when user have jwtproxy name?

Reply:
ok, I got it. Yeah, the current solution is good enough, overcomplicating it is not needed. Thanks for the clarification!